### PR TITLE
Memory–pair game

### PR DESCRIPTION
**Pull request by beta-version-profile**
_Complete the form below and remove this line._

### Memory–pair game 

[Demo](http://used-trouble.surge.sh) |
[Code base](https://github.com/beta-version-profile/frontend-2019-homeworks/tree/memory%E2%80%93pair_game/submissions/beta-version-profile/memory-pair_game)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/beta-version-profile/memory-pair_game/app.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    moveCounter()
+    if (openedCards[0].type === openedCards[1].type) {
+      matching()
+    } else {
+      notMatching()
+    }
+  }
+}
+
+for (let i = 0; i < cards.length; i++) {
```
Comment (zonzujiro) [inline submissions/beta-version-profile/memory-pair_game/app.js]: Array methods pls

File: submissions/beta-version-profile/memory-pair_game/index.css [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+  animation-name: flipInY;
+  animation-duration: 0.5s;
+}
+
+.card.show > img {
+  visibility: visible;
+}
+
+.show {
+  visibility: visible !important;
```
Comment (zonzujiro) [inline submissions/beta-version-profile/memory-pair_game/index.css]: Usually `!important` is a bad sign in CSS. It means, that you need to change normal CSS behavior.

I think you can replace this part with CSS cascade logic.

File: submissions/beta-version-profile/memory-pair_game/index.css [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+  opacity: 100 !important;
+}
+
+.card.matching {
+  cursor: default;
```
Comment (zonzujiro) [inline submissions/beta-version-profile/memory-pair_game/index.css]: Do you need this?

File: submissions/beta-version-profile/memory-pair_game/index.html [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+            <span class="moves">0</span> moves <span class="timer"></span>
+          </span>
+          <span class="right-span">
+            <span class="restart"> <i class="fas fa-redo"></i> </span>
+          </span>
+        </section>
+
+        <ul class="card-deck">
+          <li class="card" type="angel">
```
Comment (zonzujiro) [inline submissions/beta-version-profile/memory-pair_game/index.html]: `type` attribute on `li` element semanticaly needed for something [else](https://www.w3schools.com/tags/att_li_type.asp).

I think you should use `data-*` here to achieve the same result.

**Comment by beta-version-profile** [comment]
Thanks for improving my code. I learned a couple of chips ;)

**Comment by OleksiyRudenko** [comment]
@beta-version-profile Please, remove `.DS_Store` from the PR.

You may want to add a global `.gitignore` with rules for the most common IDEs etc.
"Global" means [global for your machine](https://gist.github.com/subfuzion/db7f57fff2fb6998a16c).


If you need support on this, please, let me know.

**Comment by beta-version-profile** [comment]
I done.

**Comment by OleksiyRudenko** [comment]
@beta-version-profile There is at least one more. Please, x-check your entire submission.
image

File: submissions/beta-version-profile/memory-pair_game/app.js [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    container.appendChild(cards[i])
+  }
+  openedCards = []
+
+  moves = 0
+  counter.innerHTML = moves
+
+  for (let i = 0; i < stars.length; i++) {
+    stars[i].style.color = '#ffd700'
+    stars[i].style.display = 'inline'
```
[the remark this answers is not in the document]
Comment (beta-version-profile) [reply submissions/beta-version-profile/memory-pair_game/app.js]: it's done

File: submissions/beta-version-profile/memory-pair_game/app.js [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+  disable()
+  setTimeout(() => {
+    openedCards[0].classList.remove('show', 'open', 'not-matching')
+    openedCards[1].classList.remove('show', 'open', 'not-matching')
+    enable()
+    openedCards = []
+  }, DELAY)
+}
+
+const disable = () => {
```
Comment (zonzujiro) [inline submissions/beta-version-profile/memory-pair_game/app.js]: `disable` what? 🤔 I think `disableStars` is more understandable

File: submissions/beta-version-profile/memory-pair_game/app.js [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+  cards.forEach(card => card.classList.add('disabled'))
+}
+
+const enable = () => {
+  let matchingCard = document.querySelectorAll('.matching')
+  Array.from(cards)
```
Comment (zonzujiro) [inline submissions/beta-version-profile/memory-pair_game/app.js]: Unused arrays from `cards`

File: submissions/beta-version-profile/memory-pair_game/app.js [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+  cards.filter(card => {
```
Comment (zonzujiro) [inline submissions/beta-version-profile/memory-pair_game/app.js]: Are you sure, that you need `Array#filter` here? You not filtering

File: submissions/beta-version-profile/memory-pair_game/app.js [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+      second = 0
+    }
+    if (minute == 60) {
+      hour++
+      minute = 0
+    }
+  }, 1000)
+}
+
+const congratulations = () => {
```
Comment (zonzujiro) [inline submissions/beta-version-profile/memory-pair_game/app.js]: Please use verbs for names of functions, so `showCongratulations` will be better.

File: submissions/beta-version-profile/memory-pair_game/app.js [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+  modal.classList.remove('show')
+  startGame()
+}
+
+document.body.onload = startGame()
+restartButton.addEventListener('click', startGame)
+modalPlayAgainButton.addEventListener('click', reset)
+
+function startGame() {
+  cards = shuffle(cards)
```
Comment (zonzujiro) [inline submissions/beta-version-profile/memory-pair_game/app.js]: Your array will be shuffled in place, so no need in returning the value.

File: submissions/beta-version-profile/memory-pair_game/app.js [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+let openedCards = []
+
+let second = 0,
+  minute = 0,
+  hour = 0
+
+let interval
+
+const shuffle = array => {
```
Comment (zonzujiro) [inline submissions/beta-version-profile/memory-pair_game/app.js]: FYI, you can just use `array.sort(() => 0.5 - Math.random())`

File: submissions/beta-version-profile/memory-pair_game/app.js [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+  let len = openedCards.length
+
+  if (len == 1 && moves == 0) {
+    second = 0
+    minute = 0
+    hour = 0
+    startTimer()
+  } else if (len === 2) {
+    moveCounter()
+    if (openedCards[0].dataset.type === openedCards[1].dataset.type) {
```
Comment (zonzujiro) [inline submissions/beta-version-profile/memory-pair_game/app.js]: IMO, you can use the same pattern: `const [ firstOpenedCard, secondOpenedCard ] = openedCards`

**Pull request merged**